### PR TITLE
Arc downcast

### PR DESCRIPTION
**Pull request by jsgf**
Implement `downcast` for `Arc<Any + Send + Sync>` as part of #44608, and gated by the same `rc_downcast` feature.

This PR is mostly lightly-edited cut'n'paste.

This has two additional changes:
- The `downcast` implementation needs `Any + Send + Sync` implementations for `is` and `Debug`, and I added `downcast_ref` and `downcast_mut` for completeness/consistency. (Can these be insta-stabilized?)
- At @SimonSapin's suggestion, I converted `Arc` and `Rc` to use `NonNull::cast` to avoid an `unsafe` block in each which tidied things up nicely.

### CODE REVIEW

**Comment by SimonSapin** [comment]
Stability attributes on trait impls don’t really work, so those impls are effectively insta-stable if both the trait and type are stable.

`Any + Send + Sync` methods can be made unstable and on principle should, but since they’re identical to existing `Any` and `Any + Send` methods I’m ok with landing them as insta-stable.

In both cases though, the attribute should still be accurate and say `since = "1.28.0"`. That probably requires having a different `feature` name, maybe `any_send_sync_methods` or something.

**Comment by jsgf** [comment]
OK, I made them
```
    #[stable(feature = "any_send_sync_methods", since = "1.28.0")]
```

**Comment by SimonSapin** [comment]
r=me on the diff. Let’s also get team sign off on insta-stable APIs:

* `impl Debug for Any + Send + Sync`
* `<Any+Send+Sync>::is:<T: Any>(&self) -> bool`
* `<Any+Send+Sync>::is:<T: Any>(&self) -> bool`
* `<Any+Send+Sync>::is:<T: Any>(&self) -> bool`

@bors fcp merge

**Comment by SimonSapin** [comment]
r=me on the diff. Let’s also get team sign off on insta-stable APIs:

* `impl Debug for Any + Send + Sync`
* `<Any+Send+Sync>::is:<T: Any>(&self) -> bool`
* `<Any+Send+Sync>::downcast_ref:<T: Any>(&self) -> Option<&T>`
* `<Any+Send+Sync>::downcast_mut:<T: Any>(&mut self) -> Option<&mut T>`

@bors fcp merge

**Comment by SimonSapin** [comment]
@rfcbot fcp merge

File: src/liballoc/arc.rs [hunk 1]
```diff
@@ -971,6 +972,44 @@ unsafe impl<#[may_dangle] T: ?Sized> Drop for Arc<T> {
     }
 }
 
+impl Arc<Any + Send + Sync> {
+    #[inline]
```
Comment (burtonageo) [inline src/liballoc/arc.rs]: Attributes tend to go beneath doc comments

Comment (burtonageo) [inline src/liballoc/arc.rs]: Attributes tend to go below doc comments

**Review by burtonageo — CHANGES_REQUESTED** [review]

File: src/liballoc/arc.rs (hunk 1 above)
Comment (jsgf) [reply src/liballoc/arc.rs]: This is a copy of the corresponding code in `Rc`.

**Comment by SimonSapin** [comment]
@bors r+

Thanks!

**Comment by jsgf** [comment]
I rebased onto current master (though I didn't need to resolve any conflicts).

**Comment by pthariensflame** [comment]
Should this be tagged `relnotes`?

**Comment by pietroalbini** [comment]
Yep, thanks @pthariensflame!

**Comment by SimonSapin** [comment]
For the `Any + Send + Sync` parts, yes. The `Arc` parts are still unstable.

**Pull request merged**